### DISCUSSION TITLE
How to easily transform all values in a YAML datasource (map)

### DISCUSSION THREAD

**Discussion by mattmc3** [hairyhenderson/gomplate#1604] [Q&A]
I'm using an [Alacritty YAML file](https://raw.githubusercontent.com/lighthaus-theme/alacritty/master/src/lighthaus.yml) as a datasource. It has a bunch of color values in it that I would like to transform before I can use them. Basically, I want to bulk change the '#00BFA4' style colors to '00bfa4' without having to do each and every one individually (ie: avoid running `| strings.ReplaceAll "#" "" | strings.ToLower` 25+ times).

I can get somewhat close to the results I'm hoping for, but turned my `map` into a string:

```zsh
$ url=https://raw.githubusercontent.com/lighthaus-theme/alacritty/master/src/lighthaus.yml?type=application/yaml
$ gomplate -d "alacritty=$url" -i '{{ $str := (datasource "alacritty").colors.normal | strings.ReplaceAll "#" "" | strings.ToLower }}{{ $str }}'
map[black:18191e blue:1d918b cyan:00bfa4 green:44b273 magenta:d16bb7 red:fc2929 white:cccccc yellow:e25600]
```

If I try to loop through the map, I get `bad character U+005B '['"`.

```zsh
{{ range $key, $value := $map }}
{{ $map[$key] = ( $value | strings.ReplaceAll "#" "" ) }}
{{ end }}
```

Clearly I'm a bit lost as to how to proceed.

**Comment by hairyhenderson** [comment]
> If I try to loop through the map, I get `bad character U+005B '['"`.

It's not the looping that's erroring, but your `$map[$key] =` syntax, which is invalid.

gomplate doesn't allow reassigning map values (because it's not supported by Go's templating language), so this won't work.

Instead of reassigning the value you could simply use it within that loop:

```
{{ range $key, $value := $map }}
do something with {{ strings.ReplaceAll "#" "" $value | strings.ToLower }}
{{ end }}
```

here's a working sample:

```console
$ gomplate -d 'alacritty=https://raw.githubusercontent.com/lighthaus-theme/alacritty/master/src/lighthaus.yml?type=application/yaml' \
  -i '{{ range $key, $value := (ds "alacritty").colors.normal -}}
do something with {{ strings.ReplaceAll "#" "" $value | strings.ToLower }}
{{ end }}'
do something with 18191e
do something with 1d918b
do something with 00bfa4
do something with 44b273
do something with d16bb7
do something with fc2929
do something with cccccc
do something with e25600
```

An alternative, since that particular file is otherwise in lowercase anyway (except for comments), would be to `include` it, and preprocess it:

```console
$ gomplate -d 'alacritty=https://raw.githubusercontent.com/lighthaus-theme/alacritty/master/src/lighthaus.yml?type=application/yaml' \
  -i '{{ $map := include "alacritty" | strings.ToLower | data.YAML -}}
{{ range $key, $value := $map.colors.normal -}}
do something with {{ strings.ReplaceAll "#" "" $value }}
{{ end }}'
do something with 18191e
do something with 1d918b
do something with 00bfa4
do something with 44b273
do something with d16bb7
do something with fc2929
do something with cccccc
do something with e25600
```

**Comment by mattmc3** [comment]
Thanks for the assist. That's not quite what I was after. I'm trying to transform in that loop so I can use the values in the clear later in my template rather than right then and there without having to assign each map value to its own variable. Even transforming and then assigning them to a second scrubbed map would be fine, but I don't see how I can declare new maps. Right now, I'm pre-processing with [yq](https://mikefarah.gitbook.io) because I need these values transformed before I can use them, which might be the simplest way to get what I'm after.

**Reply by hairyhenderson** [reply]
> so I can use the values in the clear later in my template

ok, so the `include` approach should work... just pipe it through a `strings.ReplaceAll`, or if you want to be a bit more selective use [`regexp.Replace`](https://docs.gomplate.ca/functions/regexp/#regexp-replace), like this:

```
{{ $map := include "alacritty" | regexp.Replace "#([0-9A-F]{6})" "$1" | strings.ToLower | data.YAML -}}
```

Then you can iterate through $map however you like.
